File: libs/domain/src/lib/domain/auth/user.model.ts
import { IResource, Resource } from '@hal-form-client';
import { Page } from '../hateoas/pageable.model';
import { Authority } from './authority.model';
import { Role } from './role.model';

export interface UserInput {
  username: string;
  email: string;
  password: string;
  firstname?: string;
  lastname?: string;
  profileImageUrl?: string;
  roleId?: string;
  authorityIds?: string[];
  active?: boolean;
  locked?: boolean;
  expired?: boolean;
  credentialsExpired?: boolean;
}

export class User extends Resource implements IResource {
  id?: string;
  email?: string;
  firstname?: string;
  lastname?: string;
  username?: string;
  profileImageUrl?: string;
  userPreferences?: UserPreferences;
  lastLoginDateDisplay?: Date;
  joinDate?: Date;
  role?: Role;
  authorities?: Authority[];
  active?: boolean;
  locked?: boolean;
  expired?: boolean;
  credentialsExpired?: boolean;
}

export interface UserUpdateProfileInput {
  firstname: string;
  lastname: string;
  profileImageUrl: string;
}

export interface UserChangePasswordInput {
  password: string;
  newPassword: string;
}

export class UserPage extends Resource implements IResource {
  page?: Page;
  override _embedded!: {
    userModels?: User[];
  };
}

export class UserPreferences extends Resource implements IResource {
  id?: string;
  darkMode?: boolean;
  contentLanguage?: string;
}

File: apps/api/src/main/java/dev/kurama/api/core/hateoas/model/UserModel.java
package dev.kurama.api.core.hateoas.model;

import lombok.*;
import org.springframework.hateoas.RepresentationModel;

import java.util.Date;
import java.util.Set;

@EqualsAndHashCode(callSuper = true)
@AllArgsConstructor
@NoArgsConstructor
@Data
@Builder
public class UserModel extends RepresentationModel<UserModel> {

  private String id;
  private String firstname;
  private String lastname;
  private String username;
  private String email;
  private String profileImageUrl;
  private Date lastLoginDateDisplay;
  private Date joinDate;
  private RoleModel role;
  private Set<AuthorityModel> authorities;
  private boolean active;
  private boolean locked;
  private boolean expired;
  private boolean credentialsExpired;

  private UserPreferencesModel userPreferences;
}
